File: pr1-java/persistence/src/main/java/pr1Java/persistence/database/UserDbRepository.java
package pr1Java.persistence.database;

import pr1Java.model.User;
import pr1Java.model.exceptions.DuplicateException;
import pr1Java.model.exceptions.NotFoundException;
import pr1Java.persistence.Configuration;
import pr1Java.persistence.UserRepository;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class UserDbRepository implements UserRepository {
    public UserDbRepository() {
        Configuration.loadProperties("./persistence/persistence.config");
        Configuration.logger.info("Initializing UserDbRepository with {} ", Configuration.properties);
    }

    @Override
    public Iterable<User> getAll() {
        Configuration.logger.traceEntry("Getting all users");

        Connection connection = DbUtils.getConnection();
        List<User> users = new ArrayList<>();
        try (ResultSet result = connection.prepareStatement("select * from users;").executeQuery()) {
            while (result.next()) {
                String username = result.getString("username");
                String password = result.getString("password");
                User user = new User(username, password);
                users.add(user);
            }
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
        }

        Configuration.logger.traceExit(users);
        return users;
    }

    @Override
    public User getOne(String username) throws NotFoundException {
        Configuration.logger.traceEntry("Getting user with {}", username);

        Connection connection = DbUtils.getConnection();
        try (ResultSet result = connection.prepareStatement("select * from users where username = '" + username + "';").executeQuery()) {
            String password = result.getString("password");
            User user = new User(username, password);

            Configuration.logger.traceExit(user);
            return user;
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
            throw new NotFoundException("user not found");
        }
    }

    @Override
    public User add(User user) throws DuplicateException {
        Configuration.logger.traceEntry("Saving {} ", user);

        try {
            getOne(user.getUsername());
            throw new DuplicateException("user already registered");
        } catch (NotFoundException ignored) {
        }

        Connection connection = DbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("insert into users(username, password) values(?,?);")) {
            preparedStatement.setString(1, user.getUsername());
            preparedStatement.setString(2, user.getPassword());

            int result = preparedStatement.executeUpdate();
            Configuration.logger.trace("Saved {} instances", result);
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
        }

        Configuration.logger.traceExit(user);
        return user;
    }

    @Override
    public User remove(String username) throws NotFoundException {
        Configuration.logger.traceEntry("Removing user with {} ", username);

        User user = getOne(username);
        Connection connection = DbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("delete from users where username = '" + username + "';")) {
            int result = preparedStatement.executeUpdate();
            Configuration.logger.trace("Deleted {} instances", result);
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
            throw new NotFoundException("user not registered");
        }

        Configuration.logger.traceExit(user);
        return user;
    }

    @Override
    public User modify(String username, User newUser) throws NotFoundException {
        Configuration.logger.traceEntry("Modifying user with {} ", username);

        User oldUser = getOne(username);
        Connection connection = DbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("update users set " +
                "password = '" + newUser.getPassword() +
                "' where username = " + username + ";")) {
            int result = preparedStatement.executeUpdate();
            Configuration.logger.trace("Modified {} instances", result);
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
        }

        Configuration.logger.traceExit(oldUser);
        return oldUser;
    }

    @Override
    public String toString() {
        StringBuilder string = new StringBuilder();
        for (User user : getAll())
            string.append(user).append("\n");
        return string.toString();
    }
}

File: pr1-cs/Persistence/Database/UserDbRepository.cs
using System.Collections.Generic;
using System.Linq;
using Model;
using Model.Exceptions;
using Mono.Data.Sqlite;

namespace Persistence.Database
{
    public class UserDbRepository : IUserRepository
    {
        public IEnumerable<User> GetAll()
        {
            var users = new List<User>();
            var connection = DbUtils.Connection;
            using var command = connection.CreateCommand();
            command.CommandText = "select * from users;";
            using var dataReader = command.ExecuteReader();
            while (dataReader.Read())
            {
                var username = dataReader.GetString(dataReader.GetOrdinal("username"));
                var password = dataReader.GetString(dataReader.GetOrdinal("password"));

                var user = new User(username, password);
                users.Add(user);
            }

            return users;
        }

        public User GetOne(string username)
        {
            User user;
            var connection = DbUtils.Connection;
            using var command = connection.CreateCommand();
            command.CommandText = "select * from users where username = '" + username + "';";
            using var dataReader = command.ExecuteReader();
            if (dataReader.Read())
            {
                var password = dataReader.GetString(dataReader.GetOrdinal("password"));

                user = new User(username, password);
            }
            else
            {
                throw new NotFoundException();
            }

            return user;
        }

        public User Add(User user)
        {
            var connection = DbUtils.Connection;
            using var command = connection.CreateCommand();
            command.CommandText = "insert into users(username, password) values(@username, @password);";

            var dataParameter = command.CreateParameter();
            dataParameter.ParameterName = "@username";
            dataParameter.Value = user.Username;
            command.Parameters.Add(dataParameter);

            dataParameter = command.CreateParameter();
            dataParameter.ParameterName = "@password";
            dataParameter.Value = user.Password;
            command.Parameters.Add(dataParameter);

            try
            {
                command.ExecuteNonQuery();
            }
            catch (SqliteException exception)
            {
                throw new DuplicateException();
            }

            return user;
        }

        public User Remove(string username)
        {
            var user = GetOne(username);
            var connection = DbUtils.Connection;
            using var command = connection.CreateCommand();
            command.CommandText = "delete from users where username = '" + username + "';";
            var result = command.ExecuteNonQuery();
            if (result == 0)
                throw new NotFoundException();
            return user;
        }

        public User Modify(string username, User newUser)
        {
            var oldUser = GetOne(username);
            var connection = DbUtils.Connection;
            using var command = connection.CreateCommand();
            command.CommandText = "update users set " +
                                  "password = '" + newUser.Password +
                                  "' where username = '" + username + "';";
            var result = command.ExecuteNonQuery();
            if (result == 0)
                throw new NotFoundException();
            return oldUser;
        }
        
        public override string ToString()
        {
            return GetAll().Aggregate("", (current, user) => current + (user + "\n"));
        }
    }
}